lib/libzpool/util.c

```diff
@@ -300,3 +302,177 @@ set_global_var(char *arg)
 
 	return (0);
 }
+
+static void
+locate_binary_path(const char *binary, char *path, size_t len)
+{
+	/*
+	 * Try to use the in-tree path. If not successful, just
+	 * allow execv to search for binary through PATH.
+	 */
+	if (realpath(getexecname(), path) == NULL)
+		goto filename;
+
+	/*
+	 * our executable basename is typically "ztest" or "zdb"
+	 */
+	char *basename = strrchr(path, '/');
+	if (basename == NULL)
+		goto filename;
+	basename++;
+
+	/*
+	 * in-tree match is typically for "/cmd/ztest/" or "/cmd/zdb/"
+	 */
+	char match[MAXNAMELEN];
+	(void) snprintf(match, sizeof (match), "/cmd/%s/", basename);
+
+	char *parent = strstr(path, match);
+	if (parent != NULL) {
+		struct stat s;
+		size_t resid = len - (parent - path);
+		(void) snprintf(parent, resid, "/cmd/%s/%s", binary, binary);
+		if (stat(path, &s) == 0)
+			return;
+	}
+
+filename:
+	strlcpy(path, binary, len);
+}
+
+static int
+run_zpool_import_config(const char *pool, const char *config, int nsearch,
+    char *searchdirs[])
+{
+	char *bin;
+	char **argv;
+	const int len = MAXPATHLEN + MAXNAMELEN + 20;
+	int result;
+	int argc = 0, argsiz;
+
+	bin = umem_alloc(len, UMEM_NOFAIL);
+	locate_binary_path("zpool", bin, len);
+
+	/*
+	 * basic argv set = {"zpool", "import", "-C", tmpfile, pool, NULL}
+	 */
+	argsiz = 6 * sizeof (char *);
+	if (nsearch > 0)
+		argsiz += (nsearch * 2) * sizeof (char *);
+	argv = umem_alloc(argsiz, UMEM_NOFAIL);
+
+	argv[argc++] = bin;
+	argv[argc++] = "import";
+	argv[argc++] = "-C";
+	argv[argc++] = (char *)config;
+	for (int i = 0; i < nsearch; i++) {
+		argv[argc++] = "-d";
+		argv[argc++] = searchdirs[i];
+	}
+	argv[argc++] = (char *)pool;
+	argv[argc++] = NULL;
+
+	pid_t pid = vfork();
+	switch (pid) {
+	case -1:
+		_exit(-1);
+
+	case 0: /* Child process */
+	{
+		int fd = open("/dev/null", O_WRONLY);
+
+		if (fd > 0) {
+			(void) dup2(fd, 1);
+			(void) dup2(fd, 2);
+			(void) close(fd);
+		}
+
+		if (bin[0] == '/')
+			(void) execvp(bin, argv);
+		else
+			(void) execv(bin, argv);
+
+		_exit(-1);
+	}
+	default: /* Parent process */
+	{
+		int error, status;
+
+		while ((error = waitpid(pid, &status, 0)) == -1 &&
+		    errno == EINTR) { }
+
+		if (error < 0 || !WIFEXITED(status)) {
+			result = -1;
+			break;
+		}
+
+		result = WEXITSTATUS(status);
+		break;
+	}
+	}
+
+	umem_free(bin, len);
+	umem_free(argv, argsiz);
+
+	return (result);
+}
+
+/*
+ * Run the zpool(8) command to find the config for 'pool'
```

Comment:
This only leaves `zhack(1)`.  You might as well get it too while you're at it, I believe this same import logic is the only reason it requires `libzfs`.

Reply:
Another good point.  Looks like `zhack(1)` also supports a '-c' (import from a cache file) option that is different from the scan a directory mode. I'll investigate what it will take to remove `zpool_tryimport` from zhack.